### ISSUE TITLE
CUDA Array Interface

### ISSUE THREAD

**Issue by scarrazza** [AleCandido/gpu#2]
If you can, please have a look at http://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html#interoperability, and check if tensorflow works too (which seems the case https://stackoverflow.com/questions/58840788/how-to-transfer-cupy-arrays-to-tensorflow)

**Comment by alecandido** [comment]
The corresponding page for CuPy I already fully read today:

https://github.com/AleCandido/gpu/blob/007bc0200c145e2bd609366a0a152821af2061b6/python/README.md?plain=1#L22-L24

It was very interesting, since it gave access to the GPU memory and reuse. Unfortunately it seems to be limited to Numba, Rmm, Pytorch, and mpi4py.
The Numba page seems to add PyArrow and ArrayViews (that I didn't know), and JAX (that I knew, but I was not aware had support for CUDA, other then XLA).

Unfortunately TensorFlow is supported, but through [DLPack](https://github.com/dmlc/dlpack), that seems to do something similar, but in memory. So you can not reuse GPU ones. And it is even [experimental](https://www.tensorflow.org/api_docs/python/tf/experimental/dlpack).

About CuPy and Numba above: it seems that the whole interoperability is limited inside the Python ecosystem, I see no why to pass control to a C library, other than make the C library making use of Python interpreter (but at that point we would run GPU through Python through C/C++, the extra layer has no real advantage...).

That's why I'm still looking into SYCL, I want to understand how can I do the operation also on ROCm. But if I can limit to CUDA maybe is just simpler.

About FPGA, I can tell you we should forget for the time being (at least from the low level point of view). It seems possible to integrate, but they will need their dedicated code...
